ACE/ace/SSL/SSL_Context.cpp

```diff
@@ -274,39 +254,8 @@ ACE_SSL_Context::set_mode (int mode)
     case ACE_SSL_Context::SSLv23:
       method = ::SSLv23_method ();
       break;
-    case ACE_SSL_Context::TLSv1_client:
-      method = ::TLSv1_client_method ();
-      break;
-    case ACE_SSL_Context::TLSv1_server:
-      method = ::TLSv1_server_method ();
-      break;
-    case ACE_SSL_Context::TLSv1:
-      method = ::TLSv1_method ();
-      break;
-#if defined(TLS1_1_VERSION) && (TLS_MAX_VERSION >= TLS1_1_VERSION)
-    case ACE_SSL_Context::TLSv1_1_client:
-      method = ::TLSv1_1_client_method ();
-      break;
-    case ACE_SSL_Context::TLSv1_1_server:
-      method = ::TLSv1_1_server_method ();
-      break;
-    case ACE_SSL_Context::TLSv1_1:
-      method = ::TLSv1_1_method ();
-      break;
-#endif
-#if defined(TLS1_2_VERSION) && (TLS_MAX_VERSION >= TLS1_2_VERSION)
-    case ACE_SSL_Context::TLSv1_2_client:
-      method = ::TLSv1_2_client_method ();
-      break;
-    case ACE_SSL_Context::TLSv1_2_server:
-      method = ::TLSv1_2_server_method ();
-      break;
-    case ACE_SSL_Context::TLSv1_2:
-      method = ::TLSv1_2_method ();
-      break;
-#endif
     default:
-      method = ::SSLv3_method ();
+      method = ::SSLv23_method ();
```

Comment:
On Dec 16 15:36 -0800, Kurt Roeckx wrote:
> >      default:
> > -      method = ::SSLv3_method ();
> > +      method = ::SSLv23_method ();
> 
> You seem to be misunderstanding something, but I'm not sure what.

Thank you @kroeckx for taking the time to explain. I was indeed
confusing both the OpenSSL and the ACE_SSL APIs.

> 
> First, that part of the documentation is removed because the SSLv2
> support has been removed.  That was a section about using the
> SSLv2 compatible client hello.  You really don't want that ...

Sure. My point in the quote was that `SSLv23_*` supports anything, while
not allowing a restriction to e.g. TLSv1.2. What I had overlooked was
the hint at `SSL_CTX_set_options()` in combination with `SSL_OP_NO_SSLv3`
and friends.

> ...

> Second, As it point out, it will support multiple versions of the
> protocol.  It's the only option that supports multiple protocols.
> Like it also point out is that if you want to restrict the
> supported protocols you should tell it to use the options.

Thanks for pointing that out (also see above).

> There is actually an upcomming new API that lets you set a minimum
> version instead of using all those settings.
> 
> This is a part that you might want to add at some point, to be
> able to disable SSLv3 for instance.  That would have as effect
> that if the other side only speaks SSLv3 that you refuse to talk
> to it.

Exactly that is my key concern: I don't want to talk to a client (or
server for that matter) that tries to force insecure protocols on me.
But isn't it all there in stable OpenSSL already? I was doing this:

```cpp
ACE_SSL_Context tlsContext;
/*
 * Calls SSL_CTX_set_options() with anything but SSL_OP_NO_TLSv1_2
 */
int ret = tlsContext.filter_versions("tlsv1.2"); // TLSv1.2 only
```

Connecting to a server using the following command line fails as
expected (client wants TLSv1, server wants TLSv1.2):

    openssl s_client -connect [ip:port] -tls1

It worked the same if I had used an ACE_SSL-based client against a
`openssl s_server`.

> Third, I don't recommend using the ciphers to select a protocol,
> it's not really meant for that.

Sure, wasn't meant to read that way. Took a shortcut and didn't
reconsider my wording.

> 
> Fouth, by default you will get TLS 1.2 if both side support it
> instead of the insecure SSLv3.  You now get the strong security
> by default.  If TLS 1.3 support is added you will also
> automatically upgraded to it.

You made the key point that I hadn't seen: Using `TLSv1_2_method()`
pins TLSv1.2 without forward compatibility. Unfortunately, the way
`ACE_SSL_Context::filter_versions()` is implemented right now, it
isn't forward-compatible, either.
So this is what the new API, you referred to above, is for?